implement assessment endpoints

albertshau opened the pull request:
Implement service endpoints that create a pipeline assessment and
table assessment for a draft.

Added a tables list to the DeltaConfig to configure which tables
the source should read from.

albertshau committed:
implement assessment endpoints

Implement service endpoints that create a pipeline assessment and
table assessment for a draft.

Introduced APIs to the DeltaSource for converting a raw list of
SQL column types into a standard CDAP Schema that will be sent
to the DeltaTarget. Also added an API for the source to assess a
table so that we can capture partial column support, such as a
db specific type that doesn't have full support in CDAP. For example,
some datetime types will lose granularity, so they are partially
supported.

Added a tables list to the DeltaConfig to configure which tables
the source should read from.

yaojiefeng commented on delta-api/src/main/java/io/cdap/delta/api/assessment/Support.java:
```diff
@@ -0,0 +1,26 @@
+package io.cdap.delta.api.assessment;
+
+/**
+ * Support level of a feature.
```
I think more javadoc is needed here about what this enum is doing.

yaojiefeng commented on delta-storage/src/main/java/io/cdap/delta/store/DraftService.java:
```diff
@@ -133,15 +151,142 @@ public TableList listDraftTables(DraftId draftId, Configurer configurer) throws
+    List<Problem> connectivityIssues = new ArrayList<>();
+    List<TableSummaryAssessment> tableAssessments = new ArrayList<>();
+    // go through all tables that the pipeline should read, fetching detail about each of the tables
+    for (SourceTable sourceTable : draft.getConfig().getTables()) {
```
Does this mean the accessment can only be done if the draft provides the specifc table names? What will happen if the draft does not specify it and wants to read from all tables?

albertshau replied:
The current mocks force the user to select tables, so there isn't a planned way to tell it to read all tables in a database. 
That being said, it's pretty straightforward to change it such that specifying nothing means reading everything, so I'll modify that.

yaojiefeng commented on delta-api/src/main/java/io/cdap/delta/api/assessment/SourceTableDetail.java:
```diff
@@ -0,0 +1,29 @@
+/**
+ * Detailed information about a source table.
+ */
+public class SourceTableDetail extends TableDetail<List<ColumnDetail>> {
```
I don't get why we need a separate class for source, seems like this class is no different than `TableDetail`, we can just create`TableDetail` right?

albertshau replied:
the difference is that the source one represents the columns differently, since we want to use the SQL names for the column types and not the CDAP Schema types. The two classes are confusing though, so I'll combine them into a single one that has both a `List<ColumnDetail>` and a `Schema`.

albertshau replied:
actually, I can't combine them because there are cases when the `List<ColumnDetail>` contains types that are not supported by the source, usually because there isn't a good representation for them in CDAP Schema. Geo objects are an example of this.

I think the confusing part is that TableDetail is generic. I'm just going to remove that class

yaojiefeng commented on delta-api/src/main/java/io/cdap/delta/api/assessment/StandardizedTableDetail.java:
```diff
@@ -0,0 +1,31 @@
+/**
+ * Standardized information about a source table.
+ */
+public class StandardizedTableDetail extends TableDetail<Schema> {
```
Same here

yaojiefeng commented on delta-storage/src/main/java/io/cdap/delta/store/DraftService.java:
```diff
@@ -133,15 +151,142 @@ public TableList listDraftTables(DraftId draftId, Configurer configurer) throws
+    return new TableAssessment(columnAssessments);
+  }
+
+  private ColumnAssessment merge(ColumnAssessment srcAssessment, ColumnAssessment targetAssessment) {
```
More javadoc help is helpful to understand what the merge is doing

yaojiefeng reviewed (commented):
Some questions

albertshau committed:
cleanup

Renamed Support to ColumnSupport and add more javadocs

Changed TableDetail back to what it was before instead of making
the columns generic. This is so TableDetail and StandardizedTableDetail
do not share a common parent, to make it more clear what each of
them represent. Added more javadocs around what each one is.

albertshau commented:
https://builds.cask.co/browse/DEL-BT6
